### PR TITLE
implementing rotate(w, theta)

### PR DESCRIPTION
**Pull request by rnd-debug**
Related to #1160

- [x] Implement rotate(w, theta, v)
- [x] Unit tests
- [x] Header in the code
- [x] Embedded docs
- Fixed an edge case in `rotationMatrix`
- Using  `config.predictable` to return an Array or a Matrix according to the type of `w` vector.

Additional request: Would it be possible to mark this PR with `hacktoberfest-accepted` tag? 😇  ([more about hacktoberfest](https://hacktoberfest.digitalocean.com/) and the [reason for the tag](https://hacktoberfest.digitalocean.com/hacktoberfest-update))

### CODE REVIEW

**Comment by josdejong** [comment]
Thanks @rnd-debug ! I'll look into your PR later (can't keep up with all issues/PR's at this moment, but no worries).

I'll add the `hacktoberfest-accepted` tag right now already 😉

File: src/function/matrix/rotate.js [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+      return config.predictable ? matrixRes.toArray() : matrixRes
+    },
+
+    'Matrix, number | BigNumber | Complex | Unit, Array | Matrix': function (w, theta, v) {
+      _validateVectorSize(w, 3)
+      return multiply(rotationMatrix(theta, v), w)
+    }
+  })
+
+  function _validateVectorSize (v, vSize) {
```
Comment (josdejong) [inline src/function/matrix/rotate.js:68]: It may be possible to unify those two util functions `_validateVectorSize ` and `_validateArraySize` into one. Right now the `_validateArraySize` does not check the number of dimensions and assumes it's a flat Array. We could use `arraySize` there, and create a single util function `_validateSize (actualSize, expectedSize)` for example.

**Review by josdejong — COMMENTED** [review]
Looks very good and well tested @rnd-debug, thanks! 

I made two small comments, one about making sure the behavior with data types is consistent with other functions.

File: src/function/matrix/rotate.js [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+     * @param {Array | Matrix} w                             Vector to rotate
+     * @param {number | BigNumber | Complex | Unit} theta    Rotation angle
+     * @param {Array | Matrix} [v]                           Rotation axis
+     * @return {Array | Matrix}                              Multiplication of the rotation matrix and w
+     */
+  return typed(name, {
+    'Array , number | BigNumber | Complex | Unit': function (w, theta) {
+      _validateArraySize(w, 2)
+      const matrixRes = multiply(rotationMatrix(theta), w)
+      return config.predictable ? matrixRes.toArray() : matrixRes
```
[the remark this answers is not in the document]
Comment (rnd-debug) [reply src/function/matrix/rotate.js]: Makes sense. Should we also update `rotationMatrix(angle, rotationAxisVector)` that systematically returns a Matrix today?

File: src/function/matrix/rotate.js (hunk 1 above)
Comment (rnd-debug) [reply src/function/matrix/rotate.js]: Did some modifications, that look kind of ugly.

File: src/function/matrix/rotate.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (josdejong) [reply src/function/matrix/rotate.js]: > Makes sense. Should we also update rotationMatrix(angle, rotationAxisVector) that systematically returns a Matrix today?

That is a good point, yes makes sense to use the information from `rotationAxisVector` when provided, and use `config.matrix` otherwise.

File: src/function/matrix/rotate.js (hunk 1 above)
Comment (josdejong) [reply src/function/matrix/rotate.js]: I like your refactoring! Is there anything you would still like to change about it?

Comment (rnd-debug) [reply src/function/matrix/rotate.js]: Triple ifs are scary :sweat_smile:

File: src/function/matrix/rotate.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (rnd-debug) [reply src/function/matrix/rotate.js]: ok, updated.

**Comment by rnd-debug** [comment]
Hmm, looks like Travis is frozen...

**Comment by josdejong** [comment]
Thanks @rnd-debug for the latest updates. Looks solid 💪

**Comment by josdejong** [comment]
The new function `rotate` is now available in `v7.6.0`. Thanks again @rnd-debug .
